File: backend/src/main/java/com/interviewer/service/ReportService.java
package com.interviewer.service;

import com.interviewer.model.*;
import com.interviewer.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class ReportService {

    @Autowired
    private SessionRepository sessionRepository;

    @Autowired
    private AnswerRepository answerRepository;

    @Autowired
    private ReportRepository reportRepository;

    @Autowired
    private QuestionRepository questionRepository;

    @Autowired
    private ProctoringService proctoringService;

    @Autowired
    private AnalysisService analysisService;

    /**
     * Generate a complete interview report
     */
    public InterviewReport generateReport(String sessionId) {
        // Check if report already exists
        Optional<InterviewReport> existing = reportRepository.findBySessionId(sessionId);
        if (existing.isPresent()) {
            return existing.get();
        }

        InterviewSession session = sessionRepository.findById(sessionId)
                .orElseThrow(() -> new RuntimeException("Session not found"));

        List<InterviewAnswer> answers = answerRepository.findBySessionId(sessionId);

        // --- Calculate Technical Score ---
        int totalScore = answers.stream().mapToInt(InterviewAnswer::getScore).sum();
        int totalMaxScore = answers.stream().mapToInt(InterviewAnswer::getMaxScore).sum();
        int technicalScore = totalMaxScore > 0 ? (int) Math.round((double) totalScore / totalMaxScore * 100) : 0;

        // --- Calculate Communication Score ---
        int communicationScore = 0;
        if (!answers.isEmpty()) {
            communicationScore = (int) answers.stream()
                    .mapToInt(a -> analysisService.calculateCommunicationScore(a.getAnswerText()))
                    .average()
                    .orElse(0);
        }

        // --- Calculate Confidence Score ---
        int confidenceScore = 0;
        if (!answers.isEmpty()) {
            confidenceScore = (int) answers.stream()
                    .mapToInt(a -> {
                        Question q = questionRepository.findById(a.getQuestionId()).orElse(null);
                        int timeLimit = (q != null) ? q.getTimeLimitSeconds() : 120;
                        return analysisService.calculateConfidenceScore(
                                a.getTimeTakenSeconds(), timeLimit, a.getAnswerText());
                    })
                    .average()
                    .orElse(0);
        }

        // --- Get Integrity Score ---
        int integrityScore = session.getIntegrityScore();

        // --- Calculate Overall Score ---
        int overallScore = (int) Math.round(
                technicalScore * 0.4 +
                communicationScore * 0.2 +
                confidenceScore * 0.2 +
                integrityScore * 0.2
        );

        // --- Identify Strengths & Weaknesses ---
        List<String> strengths = new ArrayList<>();
        List<String> weaknesses = new ArrayList<>();

        for (InterviewAnswer answer : answers) {
            double pct = answer.getMaxScore() > 0 ?
                    (double) answer.getScore() / answer.getMaxScore() * 100 : 0;
            // Extract topic from question
            String topic = extractTopic(answer.getQuestionText());
            if (pct >= 70) {
                if (!strengths.contains(topic)) strengths.add(topic);
            } else if (pct < 50) {
                if (!weaknesses.contains(topic)) weaknesses.add(topic);
            }
        }

        // --- Count correct answers ---
        long correctCount = answers.stream()
                .filter(a -> a.getMaxScore() > 0 && (double) a.getScore() / a.getMaxScore() >= 0.7)
                .count();

        // --- Get proctoring summary ---
        Map<String, Integer> proctoringSummary = proctoringService.getProctoringSummary(sessionId);

        // --- Calculate duration ---
        int durationMinutes = 0;
        if (session.getStartedAt() != null && session.getEndedAt() != null) {
            durationMinutes = (int) Duration.between(session.getStartedAt(), session.getEndedAt()).toMinutes();
        }

        // --- Build and save report ---
        InterviewReport report = new InterviewReport();
        report.setSessionId(sessionId);
        report.setUserId(session.getUserId());
        report.setRoleCategory(session.getRoleCategory());
        report.setTechnicalScore(technicalScore);
        report.setCommunicationScore(communicationScore);
        report.setConfidenceScore(confidenceScore);
        report.setIntegrityScore(integrityScore);
        report.setOverallScore(overallScore);
        report.setTotalQuestions(answers.size());
        report.setCorrectAnswers((int) correctCount);
        report.setStrengths(strengths.isEmpty() ? List.of("Keep practicing!") : strengths);
        report.setWeaknesses(weaknesses.isEmpty() ? List.of("Great job overall!") : weaknesses);
        report.setProctoringSummary(proctoringSummary);
        report.setDurationMinutes(durationMinutes);

        return reportRepository.save(report);
    }

    /**
     * Get report by session ID
     */
    public InterviewReport getReport(String sessionId) {
        return reportRepository.findBySessionId(sessionId)
                .orElseGet(() -> generateReport(sessionId));
    }

    /**
     * Get all reports for a user
     */
    public List<InterviewReport> getUserReports(String userId) {
        return reportRepository.findByUserId(userId);
    }

    /**
     * Extract topic from question text
     */
    private String extractTopic(String questionText) {
        if (questionText == null) return "General";

        String lower = questionText.toLowerCase();

        // Common Java topics
        if (lower.contains("oop") || lower.contains("object oriented") || lower.contains("polymorphism") ||
            lower.contains("inheritance") || lower.contains("encapsulation") || lower.contains("abstraction")) {
            return "OOP Concepts";
        }
        if (lower.contains("collection") || lower.contains("list") || lower.contains("map") ||
            lower.contains("set") || lower.contains("arraylist") || lower.contains("hashmap")) {
            return "Collections Framework";
        }
        if (lower.contains("thread") || lower.contains("synchron") || lower.contains("concurrent")) {
            return "Multithreading";
        }
        if (lower.contains("exception") || lower.contains("try") || lower.contains("catch") || lower.contains("throw")) {
            return "Exception Handling";
        }
        if (lower.contains("spring") || lower.contains("boot") || lower.contains("bean") || lower.contains("autowired")) {
            return "Spring Framework";
        }
        if (lower.contains("database") || lower.contains("sql") || lower.contains("query") || lower.contains("mongodb")) {
            return "Database";
        }
        if (lower.contains("rest") || lower.contains("api") || lower.contains("http") || lower.contains("endpoint")) {
            return "REST APIs";
        }
        if (lower.contains("design pattern") || lower.contains("singleton") || lower.contains("factory") || lower.contains("strategy")) {
            return "Design Patterns";
        }
        if (lower.contains("html") || lower.contains("css") || lower.contains("javascript") || lower.contains("dom")) {
            return "Web Development";
        }
        if (lower.contains("react") || lower.contains("angular") || lower.contains("vue") || lower.contains("frontend")) {
            return "Frontend Development";
        }
        if (lower.contains("node") || lower.contains("express") || lower.contains("backend")) {
            return "Backend Development";
        }
        if (lower.contains("teamwork") || lower.contains("leadership") || lower.contains("conflict") ||
            lower.contains("strength") || lower.contains("weakness") || lower.contains("tell me about")) {
            return "Behavioral / HR";
        }

        return "General Knowledge";
    }
}

File: backend/src/services/reportService.js
/**
 * reportService.js — Generates interview reports with detailed analytics.
 * Direct port of Java ReportService.
 */

const InterviewSession = require('../models/InterviewSession');
const InterviewAnswer = require('../models/InterviewAnswer');
const InterviewReport = require('../models/InterviewReport');
const Question = require('../models/Question');
const proctoringService = require('./proctoringService');
const analysisService = require('./analysisService');

/**
 * Generate a complete interview report
 */
async function generateReport(sessionId) {
  // Check if report already exists
  const existing = await InterviewReport.findOne({ sessionId });
  if (existing) {
    return existing;
  }

  const session = await InterviewSession.findById(sessionId);
  if (!session) throw new Error('Session not found');

  const answers = await InterviewAnswer.find({ sessionId });

  // --- Calculate Technical Score ---
  const totalScore = answers.reduce((sum, a) => sum + a.score, 0);
  const totalMaxScore = answers.reduce((sum, a) => sum + a.maxScore, 0);
  const technicalScore = totalMaxScore > 0 ? Math.round((totalScore / totalMaxScore) * 100) : 0;

  // --- Calculate Communication Score ---
  let communicationScore = 0;
  if (answers.length > 0) {
    const commScores = answers.map(a => analysisService.calculateCommunicationScore(a.answerText));
    communicationScore = Math.round(commScores.reduce((sum, s) => sum + s, 0) / commScores.length);
  }

  // --- Calculate Confidence Score ---
  let confidenceScore = 0;
  if (answers.length > 0) {
    const confScores = [];
    for (const a of answers) {
      const q = await Question.findById(a.questionId);
      const timeLimit = q ? q.timeLimitSeconds : 120;
      confScores.push(
        analysisService.calculateConfidenceScore(a.timeTakenSeconds, timeLimit, a.answerText)
      );
    }
    confidenceScore = Math.round(confScores.reduce((sum, s) => sum + s, 0) / confScores.length);
  }

  // --- Get Integrity Score ---
  const integrityScore = session.integrityScore;

  // --- Calculate Overall Score ---
  const overallScore = Math.round(
    technicalScore * 0.4 +
    communicationScore * 0.2 +
    confidenceScore * 0.2 +
    integrityScore * 0.2
  );

  // --- Identify Strengths & Weaknesses ---
  const strengths = [];
  const weaknesses = [];

  for (const answer of answers) {
    const pct = answer.maxScore > 0 ? (answer.score / answer.maxScore) * 100 : 0;
    const topic = extractTopic(answer.questionText);
    if (pct >= 70) {
      if (!strengths.includes(topic)) strengths.push(topic);
    } else if (pct < 50) {
      if (!weaknesses.includes(topic)) weaknesses.push(topic);
    }
  }

  // --- Count correct answers ---
  const correctCount = answers.filter(a =>
    a.maxScore > 0 && (a.score / a.maxScore) >= 0.7
  ).length;

  // --- Get proctoring summary ---
  const proctoringSummary = await proctoringService.getProctoringSummary(sessionId);

  // --- Calculate duration ---
  let durationMinutes = 0;
  if (session.startedAt && session.endedAt) {
    durationMinutes = Math.round(
      (new Date(session.endedAt) - new Date(session.startedAt)) / 60000
    );
  }

  // --- Build and save report ---
  const report = new InterviewReport({
    sessionId,
    userId: session.userId,
    roleCategory: session.roleCategory,
    technicalScore,
    communicationScore,
    confidenceScore,
    integrityScore,
    overallScore,
    totalQuestions: answers.length,
    correctAnswers: correctCount,
    strengths: strengths.length > 0 ? strengths : ['Keep practicing!'],
    weaknesses: weaknesses.length > 0 ? weaknesses : ['Great job overall!'],
    proctoringSummary,
    durationMinutes
  });

  return await report.save();
}

/**
 * Get report by session ID
 */
async function getReport(sessionId) {
  const existing = await InterviewReport.findOne({ sessionId });
  if (existing) return existing;
  return await generateReport(sessionId);
}

/**
 * Get all reports for a user
 */
async function getUserReports(userId) {
  return await InterviewReport.find({ userId });
}

/**
 * Extract topic from question text
 * Direct port of Java ReportService.extractTopic
 */
function extractTopic(questionText) {
  if (!questionText) return 'General';

  const lower = questionText.toLowerCase();

  if (['oop', 'object oriented', 'polymorphism', 'inheritance', 'encapsulation', 'abstraction']
    .some(k => lower.includes(k))) return 'OOP Concepts';

  if (['collection', 'list', 'map', 'set', 'arraylist', 'hashmap']
    .some(k => lower.includes(k))) return 'Collections Framework';

  if (['thread', 'synchron', 'concurrent']
    .some(k => lower.includes(k))) return 'Multithreading';

  if (['exception', 'try', 'catch', 'throw']
    .some(k => lower.includes(k))) return 'Exception Handling';

  if (['spring', 'boot', 'bean', 'autowired']
    .some(k => lower.includes(k))) return 'Spring Framework';

  if (['database', 'sql', 'query', 'mongodb']
    .some(k => lower.includes(k))) return 'Database';

  if (['rest', 'api', 'http', 'endpoint']
    .some(k => lower.includes(k))) return 'REST APIs';

  if (['design pattern', 'singleton', 'factory', 'strategy']
    .some(k => lower.includes(k))) return 'Design Patterns';

  if (['html', 'css', 'javascript', 'dom']
    .some(k => lower.includes(k))) return 'Web Development';

  if (['react', 'angular', 'vue', 'frontend']
    .some(k => lower.includes(k))) return 'Frontend Development';

  if (['node', 'express', 'backend']
    .some(k => lower.includes(k))) return 'Backend Development';

  if (['teamwork', 'leadership', 'conflict', 'strength', 'weakness', 'tell me about']
    .some(k => lower.includes(k))) return 'Behavioral / HR';

  return 'General Knowledge';
}

module.exports = { generateReport, getReport, getUserReports };
